craft_application/services/project.py

```diff
@@ -0,0 +1,362 @@
+            )
+        project_model = self._app.ProjectClass.from_yaml_data(
+            project, self.resolve_project_file_path()
+        )
+
+        if not project_model.adopt_info:
+            missing_fields: set[str] = set()
+            for field in self._app.mandatory_adoptable_fields:
+                if not getattr(project_model, field, None):
+                    missing_fields.add(field)
+            if missing_fields:
+                missing = ", ".join(repr(field) for field in sorted(missing_fields))
+                raise errors.CraftValidationError(
+                    f"'adopt-info' not set and required fields are missing: {missing}"
+                )
+
+        return project_model
+
+    def update_project_environment(self, info: craft_parts.ProjectInfo) -> None:
+        """Update a ProjectInfo's global environment."""
+        info.global_environment.update(
+            {
+                "CRAFT_PROJECT_VERSION": info.get_project_var("version", raw_read=True),
+            }
+        )
+
+    @property
+    @final
+    def is_rendered(self) -> bool:
+        """Whether the project has already been rendered."""
+        return self._project_model is not None
+
+    @final
+    def get(self) -> models.Project:
+        """Get the rendered project.
+
+        :returns: The project model.
+        :raises: RuntimeError if the project has not been rendered.
+        """
+        if not self._project_model:
+            raise RuntimeError("Project not rendered yet.")
+        return self._project_model
+
+    @final
+    def render_once(
+        self,
+        *,
+        platform: str | None = None,
+        build_for: str | None = None,
+    ) -> models.Project:
+        """Render the project model for this run.
+
+        This should only be called by the Application for initial setup of the project.
+        Everything else should use :py:meth:`get`.
+
+        If build_for or platform is not set, it tries to select the appropriate one.
+        Which value is selected is not guaranteed unless both a build_for and platform
+        are passed. If an application requires that each platform only build for
+        exactly one target, passing only platform will guarantee a repeatable output.
+
+        :param platform: The platform build name.
+        :param build_for: The destination architecture (or base:arch)
+        :returns: The rendered project
+        :raises: RuntimeError if the project has already been rendered.
+        """
+        if self._project_model:
+            raise RuntimeError("Project should only be rendered once.")
+
+        build_on = craft_platforms.DebianArchitecture.from_host()
+        if not platform or not build_for:
+            platforms = self.get_platforms()
+            if not platform:
+                # If we don't have a platform, select the first platform that matches
+                # our build-on and build-for. If we don't have a build-for, select the
```

Comment:
Seems like for the last sentence, one of the "build-for"s should be a "build-on"

Reply:
Yep, fixed. Thanks!